### DISCUSSION TITLE
Mysterious problem with classgraph javadoc.jar

### DISCUSSION THREAD

**Discussion by eitan-rosenberg** [classgraph/classgraph#583] [Q&A]
I am using eclipse (The latest) and as a rule I am using Maven to DL the Javadoc.

The file is downloaded from maven with no problem.

However, using it is a different story...

As can be seen, we see a preview of the class Javadoc.

![image](https://user-images.githubusercontent.com/48506054/137445988-2b9f54d9-4dcf-4016-bced-d1b5a5e45ea9.png)

But if I use the "open attached Javadoc in a browser" I am getting an Error:

java.io.FileNotFoundException: JAR entry io.github.classgraph/io/github/classgraph/ClassGraph.html not found in G:\MyMavenRepo\io\github\classgraph\classgraph\4.8.128\classgraph-4.8.128-javadoc.jar
	at java.base/sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:147)
	 	
The mystery continue if I look at the properties of the jar:

![image](https://user-images.githubusercontent.com/48506054/137446171-13ab3f79-dffa-44dc-9af6-a37ac0c9835d.png)

The jar is valid, and it opens with the browser.

I have other jars in this project and the Javadoc is shown with no problems.

Anyone having the same problem ?

Regards.

**Comment by lukehutch** [comment]
@eitan-rosenberg this is pretty interesting, thanks for reporting it...

This error

```
java.io.FileNotFoundException: JAR entry io.github.classgraph/io/github/classgraph/ClassGraph.html not found
```

seems to indicate that the browser is being directed to a path in the jar that includes the module name at the beginning of the classfile hierarchy. This is required for Java sources in modular projects.

ClassGraph is compiled as a non-modular (traditional classpath) jar -- however after the jar is built, the `pom.xml` includes a module descriptor at `/META-INF/versions/9/module-info.class` within the jar (to allow the jar to be used as a module, with maximum backwards compatibility by using this multi-version jar path, rather than just writing the descriptor to `/module-info.class`).

It seems that Eclipse detects the module descriptor, and therefore assumes that the Javadoc directory structure needs to be modular. I suspect that in general for projects that are built as modules, the Javadoc is written using the modular path structure.

I think this is probably an Eclipse bug: Eclipse needs to add a fallback to omit the module name from the beginning of the path and try loading the HTML files again, if the Javadoc cannot be found for the modular path.

I'm curious though if there is some required standard here that ClassGraph is not adhering to. Maybe I need to make a complete duplicate of the whole Javadoc tree under the module name...

**Reply by eitan-rosenberg** [reply]
@lukehutch thank you.

Meanwhile, I can use https://javadoc.io/doc/io.github.classgraph/classgraph/latest/index.html

Regards.

**Comment by lukehutch** [comment]
@eitan-rosenberg I created a duplicate of the Javadoc at the path `/io.github.classgraph` within the Javadoc jar. This should solve the problem when ClassGraph 4.8.129 is released (I don't have any features or bugfixes to release currently, so I don't know when this will be). Please let me know if it works for you, whenever that release is out. I'm glad you found javadoc.io, that's a great service!

**Reply by eitan-rosenberg** [reply]
@lukehutch thanks again.
I will try it when the time comes.
Regards.
